plugins/inputs/redis_sentinel/redis_sentinel.go

```diff
@@ -0,0 +1,445 @@
+	default:
+		return nil, fmt.Errorf("unsupported field type %v", fieldType)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("casting value %v failed: %v", value, err)
+	}
+
+	return castedValue, nil
+}
+
+func prepareFieldValues(fields map[string]string, typeMap map[string]configFieldType) (map[string]interface{}, error) {
+	preparedFields := make(map[string]interface{})
+
+	for key, val := range fields {
+		key = strings.Replace(key, "-", "_", -1)
+
+		valType, ok := typeMap[key]
+		if !ok {
+			continue
+		}
+
+		castedVal, err := castFieldValue(val, valType)
+		if err != nil {
+			return nil, err
+		}
+
+		preparedFields[key] = castedVal
+	}
+
+	return preparedFields, nil
+}
+
+// Reads stats from all configured servers accumulates stats.
+// Returns one of the errors encountered while gather stats (if any).
+func (r *RedisSentinel) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+
+	for _, client := range r.clients {
+		wg.Add(1)
+
+		go func(acc telegraf.Accumulator, client *RedisSentinelClient) {
+			defer wg.Done()
+
+			masters, err := client.gatherMasterStats(acc)
+			acc.AddError(err)
+
+			for _, master := range masters {
+				acc.AddError(client.gatherReplicaStats(acc, master))
+				acc.AddError(client.gatherSentinelStats(acc, master))
+			}
+
+			acc.AddError(client.gatherInfoStats(acc))
+		}(acc, client)
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+func (client *RedisSentinelClient) gatherInfoStats(acc telegraf.Accumulator) error {
+	infoCmd := redis.NewStringCmd("info", "all")
+	if err := client.sentinel.Process(infoCmd); err != nil {
+		return err
+	}
+
+	info, err := infoCmd.Result()
+	if err != nil {
+		return err
+	}
+
+	rdr := strings.NewReader(info)
+	infoTags, infoFields, err := convertSentinelInfoOutput(client.tags, rdr)
+	if err != nil {
+		return err
+	}
+
+	acc.AddFields(measurementSentinel, infoFields, infoTags)
+
+	return nil
+}
+
+func (client *RedisSentinelClient) gatherMasterStats(acc telegraf.Accumulator) ([]string, error) {
+	var masterNames []string
+
+	mastersCmd := redis.NewSliceCmd("sentinel", "masters")
+	if err := client.sentinel.Process(mastersCmd); err != nil {
+		return masterNames, err
+	}
+
+	masters, err := mastersCmd.Result()
+	if err != nil {
+		return masterNames, err
+	}
+
+	for _, master := range masters {
+		master, ok := master.([]interface{})
+		if !ok {
+			return masterNames, fmt.Errorf("unable to process master response")
+		}
+
+		m := toMap(master)
+
+		masterName, ok := m["name"]
+		if !ok {
+			return masterNames, fmt.Errorf("unable to resolve master name")
+		}
+
+		quorumCmd := redis.NewStringCmd("sentinel", "ckquorum", masterName)
+		quorumErr := client.sentinel.Process(quorumCmd)
+
+		sentinelMastersTags, sentinelMastersFields, err := convertSentinelMastersOutput(client.tags, m, quorumErr)
+		if err != nil {
+			return masterNames, err
+		}
+		acc.AddFields(measurementMasters, sentinelMastersFields, sentinelMastersTags)
+	}
+
+	return masterNames, nil
+}
```

Comment:
With returning the error in this function you changed the behavior. Before you _only_ skipped the failing `master`, with the current code you also drop _all_ masters that are _after_ the failing one.

I just want to make sure this is intentional!? If not you need to go back to the `acc.AddError(); continue` pattern.

Reply:
@srebhan I made this change on purpose after some thinking during refactoring. I figured that since we're iterating over the same server response we get from Sentinel (server is queried once per method) that there is not point in trying to continue processing the rest if we detect that one part of the response is malformed.

For example, if we are unable to process the name in `gatherMasterStats` for one master, it's safe to assume that the next one we go over will fail as well.

Reply:
Cool, makes a lot of sense. Can you please add a small comment in the beginning of the function with exactly that explanation so future readers will know?

Reply:
@srebhan Definitely! I added some comments in https://github.com/influxdata/telegraf/pull/10042/commits/df85e3ae72bd495560eb4b6011bfa41bd53ae6d3.